File: src/main/java/it/stanislas/kafka/delay/streamjoin/ClockKafkaKeyGenerator.java
package it.stanislas.kafka.delay.streamjoin;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

public class ClockKafkaKeyGenerator {
    private final static DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyyMMddHHmmss").withZone( ZoneId.systemDefault());

    public String now() {
        return at(Instant.now());
    }

    public String at(final Instant time) {
        final String key = FORMATTER.format(time);
        return key;
    }

}

File: src/main/kotlin/it/stanislas/kafka/delay/streamjoin/ClockKafkaKeyGenerator.kt
package it.stanislas.kafka.delay.streamjoin

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

class ClockKafkaKeyGenerator {
    fun now(): String {
        return at(Instant.now())
    }

    fun at(time: Instant): String {
        return FORMATTER.format(time)
    }

    companion object {
        private const val KEY_PATTERN = "yyyyMMddHHmmss"
        private val FORMATTER =
            DateTimeFormatter.ofPattern(KEY_PATTERN).withZone(ZoneId.systemDefault())
    }
}